File: golang_leetcode/linkedlist/206_reverse_linkedlist.go
package main

import (
	"golang_leetcode/datastruct"
)

//递归实现
func reverseList(head *datastruct.ListNode) *datastruct.ListNode {
	if head.Next == nil {
		return head
	}
	// 步骤 1: 先翻转 node 之后的链表
	newHead := reverseList(head.Next)

	// 步骤 2: 再把原 node 节点后继结点的后继结点指向 node，node 的后继节点设置为空(防止形成环)
	head.Next.Next = head
	head.Next = nil
	// 步骤 3: 返回翻转后的头结点
	return newHead
}

//迭代法实现
func reverseListByIteration(head *datastruct.ListNode) *datastruct.ListNode {
	var pre *datastruct.ListNode
	current := head
	for {
		if current == nil {
			break
		}
		next := current.Next
		current.Next = pre
		pre = current
		current = next
	}

	return pre
}

func main() {
	linkedList := datastruct.NewLinkedList()
	for i := 1; i < 100; i++ {
		linkedList.AddLast(&datastruct.ListNode{i, nil})
	}
	head := reverseListByIteration(linkedList.Head)
	//head := reverseList(linkedList.Head)
	linkedList.Head = head
	linkedList.Print()
}

File: java_leetcode/src/main/java/com/allei/_206_ReverseLinkedList.java
package com.allei;

/**
 * 链表翻转
 * https://leetcode-cn.com/problems/reverse-linked-list/
 */
public class _206_ReverseLinkedList {

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }

        @Override
        public String toString() {
            return "ListNode{" +
                    "val=" + val +
                    ", next=" + next +
                    '}';
        }

        public void println() {
            ListNode listNode = this;
            while (listNode != null) {
                System.out.print(listNode.val);
                listNode = listNode.next;
            }
            System.out.println();
        }
    }

    /**
     * 迭代法
     *
     * @param head
     * @return
     */
    public static ListNode iterateReverseList(ListNode head) {
        ListNode current = head, next, pre = null;
        while (current != null) {
            next = current.next;
            current.next = pre;
            pre = current;
            current = next;
        }
        return pre;
    }

    /**
     * 递归法
     *
     * @param head
     * @return
     */
    public static ListNode recursionReverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode newHead = recursionReverseList(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }


    public static void main(String[] args) {
        ListNode e = new ListNode(5, null);
        ListNode d = new ListNode(4, e);
        ListNode c = new ListNode(3, d);
        ListNode b = new ListNode(2, c);
        ListNode a = new ListNode(1, b);
        ListNode result = iterateReverseList(a);
        result.println();
        ListNode newResult = recursionReverseList(result);
        newResult.println();
    }
}
